proj-android/PowerAuthLibrary/src/main/java/io/getlime/security/powerauth/networking/client/JsonSerialization.java

```diff
@@ -0,0 +1,288 @@
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonParser;
+import com.google.gson.reflect.TypeToken;
+
+import java.nio.charset.Charset;
+
+import io.getlime.core.rest.model.base.request.ObjectRequest;
+import io.getlime.security.powerauth.core.ECIESCryptogram;
+import io.getlime.security.powerauth.core.ECIESEncryptor;
+import io.getlime.security.powerauth.exception.PowerAuthErrorCodes;
+import io.getlime.security.powerauth.exception.PowerAuthErrorException;
+import io.getlime.security.powerauth.rest.api.model.request.v3.EciesEncryptedRequest;
+import io.getlime.security.powerauth.rest.api.model.response.v3.EciesEncryptedResponse;
+
+/**
+ * The {@code JsonSerialization} class is helping with object to JSON serialization and
+ * with JSON to object deserialization.
+ */
+public class JsonSerialization {
+
+    /**
+    Private instance of {@link Gson} object.
+    */
+    private Gson gson;
+
+    /**
+     Private instance of {@link JsonParser}.
+     */
+    private JsonParser parser;
+
+    /**
+     Constant representing an empty object, serialized to JSON (e.g. empty curly brackets, {@code {}})
+     */
+    private static final byte[] EMPTY_OBJECT_BYTES = { 0x7B, 0x7D };
+
+
+    public JsonSerialization() {
+    }
+
+    // Generic object
+
+    /**
+     * Serializes object as is, into sequence of bytes in JSON format. If object parameter is null,
+     * then empty curly brackets are returned.
+     *
+     * @param object object to serialize
+     * @param <TRequest> type of object, to serialize
+     * @return JSON representation of object
+     */
+    @NonNull
+    public <TRequest> byte[] serializeObject(@Nullable TRequest object) {
+        if (object != null) {
+            final String jsonString = getGson().toJson(object);
+            return jsonString.getBytes(Charset.defaultCharset());
+        }
+        return EMPTY_OBJECT_BYTES;
+    }
+
+
+    /**
+     * Deserialize object from from sequence of bytes in JSON format.
+     *
+     * @param data JSON data
+     * @param type {@link TypeToken} for object to be deserialized.
+     * @param <TResponse> type of object to be deserialized.
+     * @return deserialized object
+     */
+    @NonNull
+    public <TResponse> TResponse deserializeObject(@Nullable byte[] data, @NonNull TypeToken<TResponse> type) throws JsonParseException {
+        if (data != null) {
+            final String jsonString = new String(data, Charset.defaultCharset());
+            final TResponse object = getGson().fromJson(jsonString, type.getType());
+            if (object != null) {
+                return object;
+            }
+        }
+        final String message = data == null ? "Empty response received." : "Failed to deserialize object.";
```

Comment:
This is IMO a bit hard to read.

Reply:
I've wrapped that condition to parenthesis .